docker/runtime/php/Controller.php

```diff
@@ -50,31 +50,32 @@ private function runFunction(Request $request, Response $response)
         throw new \Exception(sprintf("Function %s not exist", $this->function));
       }
       $pid = pcntl_fork();
-      if ($pid == 0) {
-        $data = $request->getBody()->getContents();
-        if ($_SERVER['HTTP_CONTENT_TYPE'] == 'application/json') {
-          $data = json_decode($data);
-        }
-        $event = (object) array(
-          'data' => $data,
-          'event-type' => $_SERVER['HTTP_EVENT_TYPE'],
-          'event-id' => $_SERVER['HTTP_EVENT_ID'],
-          'event-time' => $_SERVER['HTTP_EVENT_TIME'],
-          'event-namespace' => $_SERVER['HTTP_EVENT_NAMESPACE'],
-          'extensions' => (object) array(
-            'request' => $request,
-            'response' => $response,
-          )
-        );
-        $res = call_user_func($this->function, $event, $this->functionContext);
-        $event->extensions->response->getBody()->write($res);
-        ob_end_clean();
-        chdir($this->currentDir);
-        return $event->extensions->response;
-      } else {
-          sleep($this->timeout);
-          posix_kill($pid, SIGKILL);
-          throw new TimeoutFunctionException();
+      switch ($pid) {
+          case 0:
+              $data = $request->getBody()->getContents();
+              if ($_SERVER['HTTP_CONTENT_TYPE'] == 'application/json') {
+                  $data = json_decode($data);
+              }
+              $event = (object) array(
+                  'data' => $data,
+                  'event-type' => $_SERVER['HTTP_EVENT_TYPE'],
+                  'event-id' => $_SERVER['HTTP_EVENT_ID'],
+                  'event-time' => $_SERVER['HTTP_EVENT_TIME'],
+                  'event-namespace' => $_SERVER['HTTP_EVENT_NAMESPACE'],
+                  'extensions' => (object) array(
+                      'request' => $request,
+                      'response' => $response,
+                  )
+              );
+              $res = call_user_func($this->function, $event, $this->functionContext);
+              ob_end_clean();
+              chdir($this->currentDir);
+              return $res;
+          case -1:
+              throw new TimeoutFunctionException();
+          default:
+              sleep($this->timeout);
+              posix_kill($pid, SIGKILL);
```

Comment:
should we keep the throwing of `TimeoutFunctionException` ?

Reply:
The problem is, this was throwing a Timeout Exception even after the successful operation has concluded.  I think it's accurate to throw an exception if that's what has happened; but it doesn't seem right to do that by default... ?

How about the following... which will allow a child process to run for the length of the timeout, then signal that the process should be killed.  When the child process exits we determine if it was a result of a timeout and *then* throw a Timeout Exception...
```
default:
    $startTime = microtime(true);

    while ($pid > 0) {
        $res = pcntl_waitpid($pid, $status, WNOHANG);

        // Deal with an exited process
        if ($res == -1 || $res > 0) {
            if (pcntl_wifsignaled($status)) {
                throw new FunctionTimeoutException();
            }
            $pid = 0;
        }

        // Deal with a timeout
        $timeElapsed = microtime(true) - $startTime;
        if ($timeElapsed > $this->timeout) {
            posix_kill($pid, SIGKILL);
        }

        // Sleep for a little while to save the processor.
        usleep(10000);
    }
```

Reply:
hi @rich-jowett,

I am not quite sure of this statement

```
The problem is, this was throwing a Timeout Exception even after the successful operation has concluded.
```

Reading again the old code:

```

        ob_end_clean();
        chdir($this->currentDir);
        return $event->extensions->response;
      } else {
          sleep($this->timeout);
          posix_kill($pid, SIGKILL);
          throw new TimeoutFunctionException();
```

Here we are killing the child process just when it's over the max timeout defined and this is almost everything we need to tackle here.